File: projects/codecs/C/examples/mines-sarsa-example/SampleExperiment.c
/* 
* Copyright (C) 2008, Brian Tanner

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

#include <stdio.h>	/* for printf */
#include <math.h> /* for sqrt */
#include <rlglue/RL_glue.h> /* RL_ function prototypes and RL-Glue types */

typedef struct {
	double mean;
	double standard_dev;
} evaluation_point_t;

int whichEpisode=0;
evaluation_point_t *evaluate_agent();



void save_result_csv(evaluation_point_t *the_score[], char *fileName){
	FILE *fp;
	int i=0;

	fp=fopen(fileName, "w");
	fprintf(fp, "#Results from SampleExperiment.c.  First line is means, second line is standard deviations.\n");
	
	for(i=0;i<21;i++){
		fprintf(fp,"%.2f,",the_score[i]->mean);
	}
	fprintf(fp,"\n");
	
	for(i=0;i<21;i++){
		fprintf(fp,"%.2f,",the_score[i]->standard_dev);
	}
	fprintf(fp,"\n");

	fclose(fp);
}

void print_score(int afterEpisodes, evaluation_point_t *the_score) {
    printf("%d\t\t%.2f\t\t%.2f\n", afterEpisodes, the_score->mean, the_score->standard_dev);
}


/*
	This function will freeze the agent's policy and test it after every 25 episodes.
*/
void offline_demo(){
	int i=0;
	int j=0;
	evaluation_point_t *this_score=0;
	evaluation_point_t *statistics[21];
	
	this_score=evaluate_agent();
	print_score(0,this_score);
	statistics[0]=this_score;
	
	for(i=0;i<20;i++){
		for(j=0;j<25;j++){
			RL_episode(0);
		}
		this_score=evaluate_agent();
		print_score((i+1)*25,this_score);
		statistics[i+1]=this_score;
	}
	
	save_result_csv(statistics,"results.csv");
	
	for(i=0;i<21;i++){
		free(statistics[i]);
	}
	
}

int main(int argc, char *argv[]) {
	printf("Starting offline demo\n----------------------------\nWill alternate learning for 25 episodes, then freeze policy and evaluate for 100 episodes.\n\n");
	printf("After Episode\tMean Return\tStandard Deviation\n-------------------------------------------------------------------------\n");
	RL_init();
	offline_demo();
	
	
	printf("\nNow we will save the agent's learned value function to a file....\n");

	RL_agent_message("save_policy results.dat");

	printf("\nCalling RL_cleanup and RL_init to clear the agent's memory...\n");

	RL_cleanup();
	RL_init();


	printf("Evaluating the agent's default policy:\n\t\tMean Return\tStandardDeviation\n------------------------------------------------------\n");
	print_score(0,evaluate_agent());
	
	printf("\nLoading up the value function we saved earlier.\n");
	RL_agent_message("load_policy results.dat");

	printf("Evaluating the agent after loading the value function:\n\t\tMean Return\tStandardDeviation\n------------------------------------------------------\n");
	print_score(0,evaluate_agent());

    printf("Evaluating the agent a few times from a fixed start state of 3,3:\n\t\tMean Return\tStandardDeviation\n-------------------------------------------\n");
    RL_env_message("set-start-state 3 3");
	print_score(0,evaluate_agent());

	printf("Evaluating the agent again with the random start state:\n\t\tMean Return\tStandardDeviation\n-----------------------------------------------------\n");
    RL_env_message("set-random-start-state");
	print_score(0,evaluate_agent());


	printf("\nProgram Complete.\n");
	RL_cleanup();

	return 0;
}


evaluation_point_t *evaluate_agent(){
	int i=0;
	double sum=0;
	double sum_of_squares=0;
	double this_return=0;
	double mean;
	double variance;
	double n=100.0f;
	evaluation_point_t *eval_point=0;
	
	RL_agent_message("freeze learning");
	for(i=0;i<10;i++){
		/* We use a cutoff here in case the policy is bad
		   and will never end an episode */
		RL_episode(5000);
		this_return=RL_return();
		sum+=this_return;
		sum_of_squares+=this_return*this_return;
	}
	
	mean=sum/n;
	variance = (sum_of_squares - n*mean*mean)/(n - 1.0f);
	eval_point=(evaluation_point_t *)malloc(sizeof(evaluation_point_t));
	eval_point->mean=mean;
	eval_point->standard_dev=sqrt(variance);

	RL_agent_message("unfreeze learning");
	return eval_point;
}

File: projects/codecs/Python/examples/mines-sarsa-example/sample_experiment.py
# 
# Copyright (C) 2008, Brian Tanner
# 
#http://rl-glue-ext.googlecode.com/
#
# Licensed under the Apache License, Version 2.0 (the "License"
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# 
# 
# $Revision$
# $Date$
# $Author$
# $HeadURL$
#



import sys
import math
import rlglue.RLGlue as RLGlue

# TO USE THIS Experiment [order doesn't matter]
# NOTE: I'm assuming the Python codec is installed an is in your Python path
#   -  Start the rl_glue executable socket server on your computer
#   -  Run the SampleSarsaAgent and SampleMinesEnvironment from this or a
#   different codec (Matlab, Python, Java, C, Lisp should all be fine)
#   -  Start this environment like:
#   $> python sample_experiment.py



#
# Just do a single evaluateAgent and print it
#
def	single_evaluation():
	this_score=evaluateAgent()
	printScore(0,this_score)



print "Starting offline demo\n----------------------------\nWill alternate learning for 25 episodes, then freeze policy and evaluate for 10 episodes.\n"
print "After Episode\tMean Return\tStandard Deviation\n-------------------------------------------------------------------------"
RLGlue.RL_init()
offlineDemo()
RLGlue.RL_cleanup()

print "\nNow we will save the agent's learned value function to a file...."

RLGlue.RL_agent_message("save_policy results.dat")

print "\nCalling RL_cleanup and RL_init to clear the agent's memory..."

RLGlue.RL_cleanup()
RLGlue.RL_init()

print "Evaluating the agent's default policy:\n\t\tMean Return\tStandardDeviation\n------------------------------------------------------"
single_evaluation()

print "\nLoading up the value function we saved earlier."
RLGlue.RL_agent_message("load_policy results.dat")

print "Evaluating the agent after loading the value function:\n\t\tMean Return\tStandardDeviation\n------------------------------------------------------"
single_evaluation()

print "Telling the environment to use fixed start state of 2,3."
RLGlue.RL_env_message("set-start-state 2 3")
RLGlue.RL_start()
print "Telling the environment to print the current state to the screen."
RLGlue.RL_env_message("print-state")
print "Evaluating the agent a few times from a fixed start state of 2,3:\n\t\tMean Return\tStandardDeviation\n-------------------------------------------"
single_evaluation()

print "Evaluating the agent again with the random start state:\n\t\tMean Return\tStandardDeviation\n-----------------------------------------------------"
RLGlue.RL_env_message("set-random-start-state")
single_evaluation()


RLGlue.RL_cleanup()
print "\nProgram Complete."



#
#	This function will freeze the agent's policy and test it after every 25 episodes.
#
def offlineDemo():
	statistics=[]
	this_score=evaluateAgent()
	printScore(0,this_score)
	statistics.append(this_score)
	
	for i in range(0,20):
		for j in range(0,25):
			RLGlue.RL_episode(0)
		this_score=evaluateAgent()
		printScore((i+1)*25,this_score)
		statistics.append(this_score)
	
	saveResultToCSV(statistics,"results.csv")

def printScore(afterEpisodes, score_tuple):
	print "%d\t\t%.2f\t\t%.2f" % (afterEpisodes, score_tuple[0], score_tuple[1])

#
# Tell the agent to stop learning, then execute n episodes with his current
# policy.  Estimate the mean and variance of the return over these episodes.
#
def evaluateAgent():
	sum=0
	sum_of_squares=0
	this_return=0
	mean=0
	variance=0
	n=10
	
	RLGlue.RL_agent_message("freeze learning")
	for i in range(0,n):
		# We use a cutoff here in case the 
		#policy is bad and will never end an episode
		RLGlue.RL_episode(5000)
		this_return=RLGlue.RL_return()
		sum+=this_return
		sum_of_squares+=this_return**2
	
	mean=sum/n
	variance = (sum_of_squares - n*mean*mean)/(n - 1.0)
	standard_dev=math.sqrt(variance)

	RLGlue.RL_agent_message("unfreeze learning")
	return mean,standard_dev


def saveResultToCSV(statistics, fileName):
	theFile = open(fileName, "w")
	theFile.write("#Results from sample_experiment.py.  First line is means, second line is standard deviations.\n")

	for thisEntry in statistics:
		theFile.write("%.2f, " % thisEntry[0])
	theFile.write("\n")

	for thisEntry in statistics:
		theFile.write("%.2f, " % thisEntry[1])
	theFile.write("\n")

	theFile.close()


